certbot/lock.py

```diff
@@ -108,17 +164,101 @@ def _lock_success(self, fd):
         # the same device and inode, they're the same file.
         return stat1.st_dev == stat2.st_dev and stat1.st_ino == stat2.st_ino
 
-    def __repr__(self):
-        repr_str = '{0}({1}) <'.format(self.__class__.__name__, self._path)
-        if self._fd is None:
-            repr_str += 'released>'
-        else:
-            repr_str += 'acquired>'
-        return repr_str
-
     def release(self):
+        # type: () -> None
         """Remove, close, and release the lock file."""
+        # It is important the lock file is removed before it's released,
+        # otherwise:
+        #
+        # process A: open lock file
+        # process B: release lock file
+        # process A: lock file
+        # process A: check device and inode
+        # process B: delete file
+        # process C: open and lock a different file at the same path
         try:
-            compat.release_locked_file(self._fd, self._path)
+            os.remove(self._path)
         finally:
-            self._fd = None
+            try:
+                if self._fd:
+                    os.close(self._fd)
+            finally:
+                self._fd = None
+
+    def is_locked(self):
+        # type: () -> bool
+        """Check if lock file is currently locked.
+        :return: True if the lock file is locked
+        :rtype: bool
+        """
+        return self._fd is not None
+
+
+# Windows locking mechanism
+# ~~~~~~~~~~~~~~~~~~~~~~~~~
+class _WindowsLockMechanism(object):
+    """
+    A Windows lock file mechanism.
+    On Windows in general, access to a file is exclusive, so opening a file
+    is an effective lock. This default behavior may be modified by an administrator,
+    so we ensure correct behavior with msvcrt. Moreover on Windows a file can be removed
+    only after it has been released: so the concurrency access that may occur on POSIX
+    system is irrelevant here, leading to a quite simple code.
+    """
+
+    def __init__(self, path):
+        # type: (str) -> None
+        self._path = path
+        self._fd = None
+
+    def acquire(self):
+        """Acquire the lock"""
+        open_mode = os.O_RDWR | os.O_CREAT | os.O_TRUNC
+
+        fd = os.open(self._path, open_mode, 0o600)
+        try:
+            msvcrt.locking(fd, msvcrt.LK_NBLCK, 1)
+        except (IOError, OSError) as err:
+            # Anything except EACCES is unexpected. Raise directly the error in that case.
+            if err.errno != errno.EACCES:
+                raise
+            os.close(fd)
+            _raise_for_certbot_lock(self._path)
+
+        self._fd = fd
+
+    def release(self):
+        """Release the lock."""
+        if self._fd:
```

Comment:
This is primarily to make mypy happy, because it consider that this reference is `Optional[int]` because of its initial declaration as `None` in the constructor.

It is true here that a second `release` will not trigger an exception, while it was the case before. Do you want an Exception raised at this point in case of `not self._fd` instead ?

Reply:
I think it's probably worth it. Keeps us from writing sloppy code.

Does it satisfy mypy to put that check at the top of the method like this?
```
if self._fd is None:
   raise Something
try:
    msvcrt.locking(self._fd, msvcrt.LK_UNLCK, 1)
...
```
Not a big deal if not, but it results in slightly less nested code.

Reply:
Ok, seems to still work even when the condition is removed. So I will just let it without no further checks.